infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStoragePodInterceptor.java

```diff
@@ -0,0 +1,154 @@
+import static org.eclipse.che.workspace.infrastructure.openshift.provision.AsyncStorageProvisioner.ASYNC_STORAGE;
+
+import io.fabric8.kubernetes.api.model.DoneablePod;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.dsl.PodResource;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesInfrastructureException;
+import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
+import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftEnvironment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Checking if starting workspace configured to persisted storage and async storage pod is running.
+ * To prevent 'Multi-Attach error for volume' async storage pod will be shutdown. After deleting
+ * asynchronous storage pod procedure of starting workspace will be continued.
+ */
+public class AsyncStoragePodInterceptor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStoragePodInterceptor.class);
+  private static final int DELETE_POD_TIMEOUT_IN_MIN = 5;
+
+  private final OpenShiftClientFactory clientFactory;
+  private final String strategy;
+
+  @Inject
+  public AsyncStoragePodInterceptor(
+      @Named("che.infra.kubernetes.pvc.strategy") String strategy,
+      OpenShiftClientFactory openShiftClientFactory) {
+    this.strategy = strategy;
+    this.clientFactory = openShiftClientFactory;
+  }
+
+  public void intercept(OpenShiftEnvironment osEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    if (!"common".equals(strategy)) {
+      return;
+    }
+
+    if (isEphemeral(osEnv.getAttributes())) {
+      return;
+    }
+
+    String namespace = identity.getInfrastructureNamespace();
+    String workspaceId = identity.getWorkspaceId();
+
+    PodResource<Pod, DoneablePod> asyncStoragePodResource =
+        getAsyncStoragePodResource(namespace, workspaceId);
+
+    if (isNull(asyncStoragePodResource.get())) { // pod doesn't exist
+      return;
+    }
+
+    try {
+      deleteAsyncStoragePod(asyncStoragePodResource)
+          .get(DELETE_POD_TIMEOUT_IN_MIN, TimeUnit.MINUTES);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new InfrastructureException(
+          "Interrupted while waiting for pod removal. " + ex.getMessage());
```

Comment:
This is going to end up in the startup log facing the user I think, so IMHO we should make this more informative - what kind of pod needed to be removed?